psycopg/psycopg/_pipeline.py

```diff
@@ -197,7 +201,16 @@ def __exit__(
             else:
                 raise
         finally:
-            self._exit()
+            try:
+                self._exit()
+            except Exception as exc2:
+                # Notice that this error might be pretty irrecoverable. It
+                # happens on COPY, for insance: even if sync succeeds, exiting
+                # fails with "cannot exit pipeline mode with uncollected results"
```

Comment:
The comment is out of date per previous commit, isn't it?

Reply:
No, not really. It's still possible to put the connection in weird state, if you treat it weirdly:

```
In [1]: import psycopg
In [2]: cnn = psycopg.connect("host=localhost")

In [3]: with cnn.pipeline() as p:
   ...:     cur = cnn.cursor()
   ...:     with cur.execute("copy (select 1) to stdout") as copy:
   ...:         pass

ProgrammingError                          Traceback (most recent call last)
...
ProgrammingError: COPY cannot be used with this method; use copy() insead

During handling of the above exception, another exception occurred:

OperationalError                          Traceback (most recent call last)
...
OperationalError: cannot exit pipeline mode with uncollected results

In [4]: cnn.execute("select 1")

OperationalError                          Traceback (most recent call last)


In [5]: with cnn.pipeline() as p:
   ...:     p.sync()

error ignored exiting <psycopg.Pipeline [ACTIVE] (host=localhost database=piro) at 0x7f365874d760>: connection not in pipeline mode
error ignored exiting <psycopg.Pipeline [ACTIVE] (host=localhost database=piro) at 0x7f365874d760>: cannot send pipeline while in COPY
internal error: cannot send pipeline while in COPY
cannot exit pipeline mode with uncollected results

OperationalError                          Traceback (most recent call last)

OperationalError: connection not in pipeline mode
```

And who knows, there might be other ways too. But this is not really under our control, that's why I removed the assert for pipeline off at the end of the block: it seems blaming us rather than blaming the user (this is a 400, not a 500 :).

I considered closing the connection there instead of leaving it in the inconsistent state, but the effect is the same (likely, the connection is now unusable) and the error messages sort of point at what broke it, more than what a "connection is close" message would say afterwards.

Reply:
I see, probably an exotic usage; but can indeed happen, thanks for the clarification.

> that's why I removed the assert for pipeline off at the end of the block: it seems blaming us rather than blaming the user (this is a 400, not a 500 :).

Totally agreed; the message of 4820f361b50906e3ae147caa979020140b0fdf36 gives of good rationale for that choice.